### PR TITLE
Validate App ID

### PR DESCRIPTION
**Pull request by camstuart**
Some users including myself have generated incorrect output
by using an invalid App ID

- Updated to offer a sensible default for App ID

if the default App ID is not accepted by the user

- Updated to validate user input for App ID

Note: It is not perfectly clear what constitutes valid.
However,  this commit at least means that a
default App ID will be valid

Resolves: #8

### CODE REVIEW

File: app/index.js [hunk 1]
```diff
@@ -44,7 +44,10 @@ module.exports = generators.Base.extend({
         type: 'input',
         name: 'app_id',
         message: 'App ID',
-        default: answers => answers.name
+        validate: (input) => {
+          return (/\-|\s/.test(input)) ? 'Invalid App ID, use alphanumeric and periods only' : true;
```
Comment (ralfstx) [inline app/index.js]: This regex only tests for dashes and spaces. All other characters like `#*/&!~` etc. would slip through. If the app id is tested, it should be a positive test such as `/[a-z0-9.]+/`

Comment (cpetrov) [inline app/index.js]: In the validation message _alphanumeric_ is missing its noun, _characters_.

Comment (camstuart) [inline app/index.js]: Yes, I kept the validation 'light' because as stated in the commit message, I do not have an exact spec for what constitutes a valid App ID.

But perhaps a different approach here should be considered;

The issue I encountered was that accepting the default App ID resulted in a failure to build in the tabris-js.com online build tool. And at that time I had no idea why. It would seem that a similar experience is what lead to the creation of #8 

So the real goal is to match the validation rules used in this generator with those that are being applied to the tabris-js.com build tool.

@ralfstx, Are you able to get your hands on the exact regex being used to validate the `widget` tag's `id` attribute in `cordova/config.xml`?

I believe the rules are actually rather complex. for example, when creating App ID's in my Apple developer account, I notice that wildcards are permissible, EG: `com.tabrisjs.*`. 

But in any case, the ultimate measure of correctness would be a scenario where the derived (or validated user input) for App ID, is always be accepted by the online build tool.

In other words, the generator should use the build tool as the source of truth, after all, that is why we are validating the App ID in the first place.

Comment (ralfstx) [inline app/index.js]: I don't have a spec. The [cordova doc](https://cordova.apache.org/docs/en/latest/config_ref/index.html#widget) calls it the app's *"reverse-domain identifier"*. I've learned that, for Android, it also has to be a valid Java package name, i.e. it must not start with a number, and must not contain reserved Java keywords such as `int` as package segments.

I'd prefer the regex to be rather strict for a start. We can still adjust it on request. After all, it's a template and not everything that is permitted is also recommendable. However, covering those exceptions like the Java keywords seem to be out of scope. We don't want to maintain this list of keywords, do we? I'd suggest something simple like:

```js
/^([a-z0-9_]+\.)*[a-z][a-z0-9_]$/i
```

Does that make sense?

Comment (camstuart) [inline app/index.js]: Yes, that makes sense. I agree that a list of Java keywords is out of scope. 

Looking at the regex you have there makes me feel that a unit test is now appropriate.
I intend to make several further project contributions, so I think now is the time to add testing.

If the roadmap includes and update to yeoman v2, then this will be especially useful.

Comment (camstuart) [inline app/index.js]: I'm not sure what you mean, can you please provide a suitable error message that you would be satisfied with?

Comment (ralfstx) [inline app/index.js]: Completely agree! There's some [documentation on writing tests](http://yeoman.io/authoring/testing.html) for Yeoman generators.

Comment (ralfstx) [inline app/index.js]: I think cpetrov meant it should read "alphanumeric characters and periods".

**Comment by camstuart** [comment]
Closing, ready to open a new PR
